nucypher/blockchain/economics.py

```diff
@@ -167,6 +167,31 @@ def staking_deployment_parameters(self) -> Tuple[int, ...]:
         )
         return tuple(map(int, deploy_parameters))
 
+    def token_supply_function(self, at_period: int) -> int:
+        if at_period < 0:
+            raise ValueError("Period must be a positive integer")
+
+        with localcontext() as ctx:
+            ctx.prec = self._precision
+
+            # Eq. 3 of the mining paper
+            t = Decimal(at_period)
+            S_0 = self.erc20_initial_supply
+            i_0 = 1
+            I_0 = i_0 * S_0  # in 1/years
+            T_half = self.token_halving  # in years
+            T_half_in_days = T_half * 365
+
+            S_t = S_0 + I_0 * T_half * (1 - 2**(-t / T_half_in_days)) / LOG2
```

Comment:
Of course, this is assuming that everyone stakes for the same time (where `T_half = 2 years` for all being 1+ year stakers)